warehouse/query-core/src/main/java/datawave/query/transformer/GroupingTransform.java

```diff
@@ -157,17 +158,34 @@ private Key createAttrKey(String fieldName) {
      *            an iterator source
      * @return the flushed value that is an aggregation from the source iterator
      */
-    public Iterator<Entry<Key,Document>> getGroupingIterator(final Iterator<Entry<Key,Document>> in, int max) {
+    public Iterator<Entry<Key,Document>> getGroupingIterator(final Iterator<Entry<Key,Document>> in, int groupFieldsBatchSize, YieldCallback yieldCallback) {
         
         return new Iterator<Entry<Key,Document>>() {
             
             Entry<Key,Document> next;
             
             @Override
             public boolean hasNext() {
-                for (int i = 0; i < max && in.hasNext(); i++) {
-                    Entry<Key,Document> lastEntry = in.next();
-                    GroupingTransform.this.apply(lastEntry);
+                for (int i = 0; i < groupFieldsBatchSize; i++) {
+                    if (in.hasNext()) {
+                        Entry<Key,Document> lastEntry = in.next();
+                        if (lastEntry != null) {
+                            GroupingTransform.this.apply(lastEntry);
+                        } else if (yieldCallback != null && yieldCallback.hasYielded()) {
+                            log.trace("lastEntry is null and yield was called");
+                            yieldCallback.yield(null); // undo the yield
```

Comment:
Don't we need to break here?  Otherwise we will continue looping and calling in.hasNext() again which is wasteful.

Reply:
This is similar to the comment below, however we may not ever have a case where in.hasNext() was true, but in.next() returns null. In any event, this was intended to assume that if something happened because of a yield, the right thing to do is to turn off the yield and see if there is another record. I believe we would only waste time here if in.hasNext() returns true over and over while in.next() returns null.

Reply:
Apparently I was not clear in my explanation so my apologies.
The reason we yielded in the first place was that this scan thread needs to return control to the Tablet.  There are two ways to return control: 1) yield, 2) return a key.  What we need to happen here is since we detected a yield, we need to cancel the yield and return a key instead.  The "last key processed" in this case needs to be the key that was registered in the yield callback before setting it to null.  Return that key along with your grouping information gathered thus far.